# Word Importance Classification

This code is main edited by Enze Wang

# Attention

Please put four .csv file in data folder and put this code in main folder instead of code folder.

This code want to provide you some important word list and classify our word. 

In [18]:
import pandas as pd
import matplotlib.pyplot as plt
import random
import numpy as np
import nltk
import string
nltk.download('stopwords')
nltk.download('punkt')
from nltk.corpus import stopwords
from nltk.stem.wordnet import WordNetLemmatizer
lmtzr = WordNetLemmatizer()
import itertools
from collections import Counter
from nltk.sentiment.vader import SentimentIntensityAnalyzer
nltk.download('vader_lexicon')
from nltk import FreqDist
from nltk.stem import PorterStemmer
nltk.download('wordnet')
from nltk.corpus import wordnet as wn
nltk.download('averaged_perceptron_tagger')
from wordcloud import WordCloud
from nltk import sent_tokenize, word_tokenize, PorterStemmer
from nltk.corpus import stopwords
from nltk.text import TextCollection
from nltk.corpus import wordnet

[nltk_data] Downloading package stopwords to /Users/enze/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package punkt to /Users/enze/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package vader_lexicon to
[nltk_data]     /Users/enze/nltk_data...
[nltk_data]   Package vader_lexicon is already up-to-date!
[nltk_data] Downloading package wordnet to /Users/enze/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!
[nltk_data] Downloading package averaged_perceptron_tagger to
[nltk_data]     /Users/enze/nltk_data...
[nltk_data]   Package averaged_perceptron_tagger is already up-to-
[nltk_data]       date!


# Word Importance with TF-IDF

This code calculate word importance by TF-IDF, another code by Yicen is by tree regression.

TF-IDF includes two parts:

TF(WORD) = (Number of times term WORD appears in a document) / (Total number of terms in the document)

IDF(WORD) = ln(Total number of documents / Number of documents with term WORD in it)

TF-IDF=TF*IDF

Therefore high TF-IDF means that this word appears many times but not as many times as words like "and" and "is" because they are too common. 

TF-IDF only consider the word frequency but has no connection with star rating. 

For example, there are three documents:

"I have an apple"

"I have a pen"

"apple pen pen apple"


In [19]:
text_all=[["I","have","an","apple"],["I","have","a","pen"],["apple","pen","pen","apple"]]
text_collection = TextCollection(text_all)
for word in ["I","have","an","apple"]:
    print(word, "idf=", text_collection.idf(word))

I idf= 0.4054651081081644
have idf= 0.4054651081081644
an idf= 1.0986122886681098
apple idf= 0.4054651081081644


"I" occur in 2 documents, and there are 3 documents, therefore its idf is ln(3/2)=0.4054651

In [20]:
all_word=[w for text in text_all for w in text]

In [21]:
for word in ["I","have","an","apple"]:
    word_tf = text_collection.tf(word, all_word)
    print(word, "tf=", word_tf)

I tf= 0.16666666666666666
have tf= 0.16666666666666666
an tf= 0.08333333333333333
apple tf= 0.25


"I" occur 2 times in 12 in the whole document, therefore its tf is 1/6=0.1666667

In [22]:
for word in ["I","have","an","apple"]:
    word_tf_idf = text_collection.tf_idf(word, all_word)
    print(word, "tf_idf=", word_tf_idf)

I tf_idf= 0.06757751801802739
have tf_idf= 0.06757751801802739
an tf_idf= 0.0915510240556758
apple tf_idf= 0.1013662770270411


So "I" tf-idf is 0.4054651081081644*0.16666666666666666=0.06757752

In [23]:
clean_review=pd.read_csv("data/pizza_business_clean_review.csv")

In [24]:
clean_review

Unnamed: 0               review_id  \
0               0  hfXVxrlD6fr1kXfz6sR13Q   
1               1  EDQpLLFiAxeSxWI5ipwzBQ   
2               2  RWmT90rP7wYo57dG1cyvww   
3               3  0jduztItUhOh3FzKGjMxZQ   
4               4  sxfmzhMvZ8LlI1x4fhNjKg   
...           ...                     ...   
67164       67164  gcNRaJMDlSVs8YlqwpsRuw   
67165       67165  yFxDOdiDb8HPaK1io81kiw   
67166       67166  Z_xmoLgIMB4As90t0a5eBQ   
67167       67167  CzYRFCnkbqOVb3WIMMrbfw   
67168       67168  ZhUeFHPAjUbEKRnYO_pl_w   

                                            clean_review  
0      parking tough get get go much also get little ...  
1      server extensive knowledge wine various menu s...  
2      fast forward every trip nyc pitt includes visi...  
3      far favorite cleveland/akron restaurant great ...  
4      food good usually get italian sub pizza would ...  
...                                                  ...  
67164  originally side ohio always really enjoyed jet...  
67165  yes yes yes loved going place shadyside ecstat...  
67166  went lunch meeting amidst typical suburban str...  
67167  wife eaten la brioche least dozen time food ne...  
67168  first let say lindy columbus one favorite rest...  

[67169 rows x 3 columns]

In [25]:
text_all=[]
r,c=clean_review.shape
for i in range(r):
    text_all.append(list(map(str,clean_review["clean_review"][i].split(" "))))
    

In [26]:
all_word=[w for text in text_all for w in text]

In [27]:
important_word=pd.read_csv("data/Word_Freq.csv")

In [28]:
important_word

Unnamed: 0          Word  Frequency
0              0         pizza      98236
1              1          good      40352
2              2         place      37016
3              3          food      35068
4              4         great      29869
...          ...           ...        ...
5510        5510       dragged         31
5511        5511    tenderloin         31
5512        5512       marotta         31
5513        5513  greenpointer         31
5514        5514       lengthy         31

[5515 rows x 3 columns]

In [29]:
Word_list=important_word["Word"].tolist()
text_collection = TextCollection(text_all)
TF_IDF=[]
for word in Word_list:
    word_tf_idf = text_collection.tf_idf(word, all_word)
    TF_IDF.append(word_tf_idf)
important_word["tf_idf"]=TF_IDF

In [30]:
important_word=important_word.sort_values(by=['tf_idf'],ascending=False)
important_word[0:20]

Unnamed: 0     Word  Frequency    tf_idf
0            0    pizza      98236  0.015003
1            1     good      40352  0.011403
3            3     food      35068  0.010822
2            2    place      37016  0.010805
5            5     time      27434  0.010486
7            7    order      21401  0.010308
4            4    great      29869  0.010277
6            6     like      22265  0.009689
8            8      one      20728  0.009455
11          11    would      19830  0.009316
10          10      get      19869  0.009151
16          16   cheese      16251  0.008807
14          14    sauce      16660  0.008790
12          12  ordered      17633  0.008710
17          17   really      16040  0.008423
9            9  service      20093  0.008378
13          13     back      17540  0.008224
20          20    salad      13206  0.008117
15          15       go      16263  0.008016
18          18     also      14279  0.007700

As you can see, there is totally no difference between the rank with frequence and the rank with TF-IDF. The reason is that the stop words has been removed from the steps above.


In [32]:
importance_word=important_word.drop(["Unnamed: 0"],axis=1)
importance_word.to_csv("data/Importance_word_tf_idf.csv")